sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientMetrics.scala

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.spark.diagnostics.BasicLoggingTrait
+import com.codahale.metrics.{MetricRegistry, Slf4jReporter}
+import io.micrometer.core.instrument.composite.CompositeMeterRegistry
+import io.micrometer.core.instrument.config.NamingConvention
+import io.micrometer.core.instrument.{Clock, MeterRegistry}
+import io.micrometer.core.instrument.dropwizard.{DropwizardConfig, DropwizardMeterRegistry}
+import io.micrometer.core.instrument.util.HierarchicalNameMapper
+
+import java.util.concurrent.TimeUnit
+
+// scalastyle:off underscore.import
+import scala.collection.JavaConverters._
+// scalastyle:on underscore.import
+
+private[spark] object CosmosClientMetrics extends BasicLoggingTrait {
+  var meterRegistry: Option[CompositeMeterRegistry] = None
+  var executorId: Option[String] = None
+  var slf4JReporter : Option[Slf4jReporter] = None
+
+  def addMeterRegistry
+  (
+    newMeterRegistry: MeterRegistry
+  ) : Unit = {
+    this.meterRegistry match {
+      case Some(existingRegistry) => existingRegistry.add(newMeterRegistry)
+      case None =>
+        this.meterRegistry = Some(new CompositeMeterRegistry(
+          Clock.SYSTEM,
+          Iterable.apply(newMeterRegistry).asJava))
+
+    }
+  }
+
+  def removeMeterRegistry
+  (
+    toBeRemovedMeterRegistry: MeterRegistry
+  ) : Unit = {
+    val shouldShutdown = this.meterRegistry match {
+      case Some(existingRegistry) =>
+        if (existingRegistry.remove(toBeRemovedMeterRegistry).getRegistries().size() == 0) {
+          true
+        } else {
+          false
```

Comment:
when it return false? and the registry will never got shutdown?

Reply:
this.meterRegistry is a CompositeMeterRegsitry - basically a delegating wrapper around a collection of MeterRegistries. If it isn't empty you can't shut the CompoositeMeterRegistry down because it still contains active meter registries

For example when you have two spark plugins registering different meter registries - and only one gets removed

Reply:
ah gotcha